lib/remap-lang/src/parser.rs

```diff
@@ -4,354 +4,363 @@ use crate::{
 
-// Auto-generate a set of parser functions to parse different operations.
-macro_rules! operation_fns {
-    (@impl $($rule:tt => { op: [$head_op:path, $($tail_op:path),+ $(,)?], next: $next:tt, })+) => (
-        $(
-            paste::paste! {
-                fn [<$rule _from_pairs>](mut pairs: Pairs<R>) -> Result<Expr> {
-                    let inner = pairs.next().ok_or(e(R::$rule))?.into_inner();
-                    let mut lhs = [<$next _from_pairs>](inner)?;
-                    let mut op = Operator::$head_op;
+        match field.as_rule() {
+            R::string => {
+                let string = field.into_inner().next().ok_or(e(R::string))?;
+                self.escaped_string_from_pair(string)
+            }
+            R::ident => Ok(field.as_str().to_owned()),
+            _ => Err(e(Rule::path_field)),
+        }
+    }
 
-                    for pair in pairs {
-                        match pair.as_rule() {
-                            R::[<operator_ $rule>] => {
-                                op = Operator::from_str(pair.as_str()).map_err(|_| e(R::$rule))?;
-                            }
-                            _ => {
-                                lhs = Expr::from(Arithmetic::new(
-                                    Box::new(lhs),
-                                    Box::new([<$next _from_pairs>](pair.into_inner())?),
-                                    op.clone(),
-                                ));
-                            }
-                        }
-                    }
+    fn path_coalesce_from_pair(&self, pair: Pair<R>) -> Result<Vec<String>> {
+        pair.into_inner()
+            .map(|pair| self.path_field_from_pair(pair))
+            .collect::<Result<_>>()
+    }
 
-                    Ok(lhs)
+    fn escaped_string_from_pair(&self, pair: Pair<R>) -> Result<String> {
+        // This is only executed once per string at parse time, and so I'm not
+        // losing sleep over the reallocation. However, if we want to mutate the
+        // underlying string then we can take some inspiration from:
+        //
+        // https://github.com/rust-lang/rust/blob/master/src/librustc_lexer/src/unescape.rs
+
+        let literal_str = pair.as_str();
+        let mut escaped_chars: Vec<char> = Vec::with_capacity(literal_str.len());
+
+        let mut is_escaped = false;
+        for c in literal_str.chars() {
+            if is_escaped {
+                match c {
+                    '\\' => escaped_chars.push(c),
+                    'n' => escaped_chars.push('\n'),
+                    't' => escaped_chars.push('\t'),
+                    '"' => escaped_chars.push('"'),
+                    // This isn't reachable currently due to the explicit list of
+                    // allowed escape chars in our parser grammar. However, if that
+                    // changes then we might need to rely on this error.
+                    _ => return Err(Error::EscapeChar(c)),
```

Comment:
Could this simply default to: `_ => escaped_chars.push(c),`?

Reply:
It seems like it could, but I'll double-check (this code was copy/pasted from the original impl).

Reply:
One reason I could see this being confusing, is if we add an escape char such as `r` to the parser, we'd have to explicitly convert that to `\r` here, so just pushing the char as a default case won't work in that case.

I think having an explicit error makes it more clear that you need to special case the new escape character.

As the comment says, there's no way this can trigger unless the PEG is updated, so I'm inclined to leave this as a more explicit case that needs to be updated if the grammar changes.

Reply:
Having said that, I realized this error doesn't have to be special-cased, compared to other "invalid token" errors.

So I've updated this a bit in adcd42fe7 (keeping the original intent of returning an error if the grammar invalidates this parse function).